llvm/lib/Transforms/Yk/LivenessAnalysis.cpp

```diff
@@ -0,0 +1,161 @@
+//===- LivenessAnalysis.cpp ------------------===//
+//
+// A liveness analysis for LLVM IR.
+//
+// This is based on the algorithm shown in Chapter 10 of the book:
+//
+//   Modern Compiler Implementation in Java (2nd edition)
+//   by Andrew W. Appel
+
+#include "llvm/Transforms/Yk/LivenessAnalysis.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+/// Wrapper to make `std::set_difference` more concise.
+///
+/// Store the difference between `S1` and `S2` into `Into`.
+void vset_difference(const std::set<Value *> &S1, const std::set<Value *> &S2,
+                     std::set<Value *> &Into) {
+  std::set_difference(S1.begin(), S1.end(), S2.begin(), S2.end(),
+                      std::inserter(Into, Into.begin()));
+}
+
+/// Wrapper to make `std::set_union` more concise.
+///
+/// Store the union of `S1` and `S2` into `Into`.
+void vset_union(const std::set<Value *> &S1, const std::set<Value *> &S2,
+                std::set<Value *> &Into) {
+  std::set_union(S1.begin(), S1.end(), S2.begin(), S2.end(),
+                 std::inserter(Into, Into.begin()));
+}
+
+namespace llvm {
+/// Find the successor instructions of the specified instruction.
+std::set<Instruction *>
+LivenessAnalysis::getSuccessorInstructions(Instruction *I) {
+  Instruction *Term = I->getParent()->getTerminator();
+  std::set<Instruction *> SuccInsts;
+  if (I != Term) {
+    // Non-terminating instruction: the sole successor instruction is the
+    // next instruction in the block.
+    SuccInsts.insert(I->getNextNode());
+  } else {
+    // Terminating instruction: successor instructions are the first
+    // instructions of all successor blocks.
+    for (unsigned SuccIdx = 0; SuccIdx < Term->getNumSuccessors(); SuccIdx++)
+      SuccInsts.insert(&*Term->getSuccessor(SuccIdx)->begin());
+  }
+  return SuccInsts;
+}
+
+/// Replaces the value set behind the pointer `S` with the value set `R` and
+/// returns whether the set behind `S` changed.
+bool LivenessAnalysis::updateValueSet(std::set<Value *> *S,
```

Comment:
This method name (indeed the entire method!) feel weird. Can we find something more descriptive?

Reply:
Maybe, but why is it weird? It updates a `set` of `Value *`, hence `updateValueSet()`.

Is the confusion in `update`? How about `replaceValueSet()`?

(Note that this is not new code)

Reply:
It's the "Replaces the value set behind the pointer `S`" bit -- it feels rather non-idiomatic. Maybe there's no other way, but it just feels... weird.

Reply:
Let me see if I can remove the need for this oddity.

Reply:
Does [a3a39a5](https://github.com/ykjit/ykllvm/pull/38/commits/a3a39a573127caf986087e883d5f77b27e6d6d84) work better?

Reply:
Apart from the disappearing doc comments, this does feel a bit nicer.